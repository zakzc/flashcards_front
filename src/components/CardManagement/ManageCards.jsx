import React, { Component } from "react";
import { Redirect } from "react-router-dom";
// hooks
import updateCards from "../../Data/Data_Update/updateCards";
import { CheckForInvalidCharacters } from "../../Data/Validation/validate";
// Icons
import Consolidate from "../icons/consolidate";
import EditCardIcon from "../icons/cardEdit";

export default class ManageCards extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userIsLoggedIn: props.userIsLoggedIn,
      user: props.currentUser,
      currentStack: props.currentStack,
      token: props.token,
      cardForEditing: "",
      cardForEditingId: "",
      editingMode: false,
      tempNewFront: "",
      tempNewBack: "",
      newStack: "",
      updatedItemMessage: false,
      redirect: false,
      messageToUser:
        "If you only want to edit one element (front OR back), leave the other element empty",
      messageToUserBack:
        "After you change the element locally, you need to click Consolidate changes button in order to reflect these changes in the actual stack.",
    };
    // utils
    this.CheckForInvalidCharacters = CheckForInvalidCharacters;
    // this.manageCurrentStack = this.manageCurrentStack.bind(this);
    this.editCard = this.editCard.bind(this);
    this.cardsInCurrentStack = this.cardsInCurrentStack.bind(this);
    this.frontEditHandler = this.frontEditHandler.bind(this);
    this.backEditHandler = this.backEditHandler.bind(this);
    this.editButtonHandler = this.editButtonHandler.bind(this);
    this.validateInput = this.validateInput.bind(this);
    this.makeNewCardSet = this.makeNewCardSet.bind(this);
    this.chooseCard = this.chooseCard.bind(this);
    this.stackWasUpdated = this.stackWasUpdated.bind(this);
    this.plusButton = this.editButton.bind(this);
    this.consolidateButton = this.consolidateButton.bind(this);
    this.consolidateButtonView = this.consolidateButtonView.bind(this);
    this.inputForm = this.inputForm.bind(this);
    this.showCurrentStack = this.showCurrentStack.bind(this);
    this.manageCardsView = this.manageCardsView.bind(this);
    this.manageCardsLayout = this.manageCardsLayout.bind(this);
  }

  ////////////////
  // Implementation methods
  ////////////////

  async handleSubmitReady() {
    let updateCardsProcess;
    updateCardsProcess = await updateCards(
      this.state.newStack,
      this.state.token
    )
      .then(() => {
        this.setState({
          messageToUser:
            "Consolidating changes and sending the new data to your database.",
          messageToUserBack: "You will be redirected",
        });
      })
      .then(() => {
        this.setState({ redirect: true });
      })
      .catch((err) => {
        this.setState({
          messageToUser: "Error on updating process (error 29)",
        });
      });
    if (!updateCardsProcess) {
      updateCardsProcess = false;
      return updateCardsProcess;
    }
  }

  frontEditHandler(e) {
    // e.preventDefault();
    this.setState({ tempNewFront: e.target.value });
  }

  backEditHandler(f) {
    // e.preventDefault();
    this.setState({ tempNewBack: f.target.value });
  }

  // gets the new values for front and back
  editButtonHandler(e) {
    e.preventDefault();
    let newValueFront;
    if (this.state.tempNewFront) {
      newValueFront = this.state.tempNewFront;
    } else {
      newValueFront = this.state.currentStack.cards[this.state.cardForEditing]
        .front;
    }
    let newValueBack;
    if (this.state.tempoNewBack) {
      newValueBack = this.state.tempNewBack;
    } else {
      newValueBack = this.state.currentStack.cards[this.state.cardForEditing]
        .back;
    }
    this.makeNewCardSet(newValueFront, newValueBack);
  }

  validateInput(front, back) {
    if (!this.CheckForInvalidCharacters(front)) {
      this.setState({
        messageToUser: "You entered an invalid character in the front card",
      });
      return false;
    }
    if (!this.CheckForInvalidCharacters(back)) {
      this.setState({
        messageToUser: "You entered an invalid character in the back card",
      });
      return false;
    }
    this.setState({
      messageToUser: "Valid input",
    });
    return true;
  }

  makeNewCardSet(newFront, newBack) {
    if (this.validateInput(newFront, newBack) === true) {
      // makes a new version of the cardSet
      let updatedStack = this.state.currentStack;
      if (newBack) {
        updatedStack.cards[this.state.cardForEditing].back = newBack;
      } else {
        updatedStack.cards[
          this.state.cardForEditing
        ].back = this.state.currentStack.cards[this.state.cardForEditing].back;
      }
      if (newFront) {
        updatedStack.cards[this.state.cardForEditing].front = newFront;
      } else {
        updatedStack.cards[
          this.state.cardForEditing
        ].front = this.state.currentStack.cards[
          this.state.cardForEditing
        ].front;
      }
      // updates local state
      this.setState((state) => ({
        newStack: updatedStack,
        editingMode: false,
        updatedItemMessage: true,
      }));
    } else {
      this.setState({ messageToUser: "invalid input" });
    }
  }

  chooseCard(e) {
    let selectedCard = e.target.value;
    this.setState(() => ({
      cardForEditing: selectedCard,
      editingMode: true,
    }));
    this.setState({ cardForEditingId: selectedCard });
  }

  ////////////////
  // View methods
  ////////////////

  infoCard() {
    return (
      <div className="flipInfoCard">
        <div className="flipInfoCardInner">
          <p className="flipInfoCardFront">{this.state.messageToUser}</p>
          <p className="flipInfoCardBack">{this.state.messageToUserBack}</p>
        </div>
      </div>
    );
  }

  consolidateButton() {
    return (
      <div id="flipButton_Positioning">
        <div className="flipContainer">
          <div className="flipInner">
            <button
              className="buttonStyle"
              type="button"
              onClick={this.handleSubmitReady}
            >
              <div className="flipFront">
                <Consolidate />
              </div>
              <div className="flipBack">
                <span className="buttonMessage">Consolidate changes</span>
              </div>
            </button>
          </div>
        </div>
      </div>
    );
  }

  editButton() {
    return (
      <button
        className="buttonStyle"
        type="button"
        onClick={this.editButtonHandler}
      >
        <div className="flipContainer">
          <div className="flipInner">
            <div className="flipFront">
              <EditCardIcon />
            </div>
            <div className="flipBack">
              <span className="buttonMessage">Change</span>
            </div>{" "}
          </div>
        </div>
      </button>
    );
  }

  inputForm() {
    return (
      <form id="inputFormCardEdit">
        <p>
          The current card Front is{" "}
          <strong className="detach">
            {this.state.currentStack.cards[this.state.cardForEditing].front}
          </strong>
        </p>
        <p>
          Inform new card <strong className="emphasis">front</strong>
        </p>
        <input type="text" name="newFront" onChange={this.frontEditHandler} />
        <p>
          The current card back is{" "}
          <strong className="detach">
            {this.state.currentStack.cards[this.state.cardForEditing].back}
          </strong>
        </p>
        <p>
          Inform new card <strong className="emphasis">back</strong>
        </p>
        <input type="text" name="newBack" onChange={this.backEditHandler} />
        <br />
        {this.editButton()}
      </form>
    );
  }

  editCard() {
    return (
      <React.Fragment>
        <div id="editCardArea">
          {this.inputForm()}
          {this.infoCard()}
        </div>
      </React.Fragment>
    );
  }

  stackWasUpdated() {
    return (
      <div id="shortMessage" className="infoCard_cardSet">
        <p>
          Card was updated locally. Select another card to update or consolidate
          changes by selecting any card and clicking on the consolidate button.
        </p>
      </div>
    );
  }

  cardsInCurrentStack() {
    return (
      <React.Fragment>
        <h4>Now, select the card you want to manage</h4>
        <div>
          <select className="select-css" onChange={this.chooseCard}>
            <option value="0">Select card</option>
            {this.state.currentStack.cards.map((c, i) => {
              return (
                <option key={i} value={i}>
                  {c.front} - {c.back}
                </option>
              );
            })}
          </select>
        </div>
      </React.Fragment>
    );
  }

  showCurrentStack() {
    return (
      <React.Fragment>
        <div>
          <h3>
            Your current stack is:{" "}
            <strong className="emphasis">
              {this.state.currentStack.stackName}
            </strong>
          </h3>
        </div>
      </React.Fragment>
    );
  }

  consolidateButtonView() {
    return (
      <div id="consolidateButton">
        {this.state.updatedItemMessage ? this.consolidateButton() : <div></div>}
      </div>
    );
  }

  manageCardsLayout() {
    return (
      <div id="manageCardsWrapper">
        <h1>Let's manage your cards</h1>

        <div id="manageCardsArea">
          {this.showCurrentStack()}
          {this.cardsInCurrentStack()}
          {this.state.updatedItemMessage ? this.stackWasUpdated() : <div></div>}
          {this.state.editingMode ? this.editCard() : <div></div>}
          {this.consolidateButtonView()}
          {this.state.redirect ? (
            <Redirect
              to="/consolidateChanges"
              userIsLoggedIn={this.state.userIsLoggedIn}
            />
          ) : (
            <div></div>
          )}
        </div>
      </div>
    );
  }

  manageCardsView() {
    // console.log("MANAGE CARDS");
    return (
      <div id="manageCardsLayout">
        {this.state.userIsLoggedIn ? (
          <div id="loggedInView">{this.manageCardsLayout()}</div>
        ) : (
          <Redirect to="/login" />
        )}
      </div>
    );
  }

  render() {
    return <React.Fragment>{this.manageCardsView()}</React.Fragment>;
  }
}
