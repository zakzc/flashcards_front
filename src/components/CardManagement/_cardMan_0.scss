
#cardManagementLayout {
  background-color: $mainBackColor;
}

#cardManagementView {
  background-color: $mainBackColor;
}


#manageCardsLayout {
  // adjust of the container positioning
  margin-top: 1%;
  margin-left: 1%;
}

#cardManagementWrapper {
  background-color: $mainBackColor;
  // color: $colorCardText;
  height: auto;
  display: flex;
  padding-bottom: 15%;
}

#editCardArea {
  margin-top: 1%;
}

#manageCardsArea {
  float: left;
  width: 90%;
}

#consolidateButton {
  float: right;
}

#shortMessage {
  color: $backColorNav;
}

input {
  width: 75%;
  padding: 12px 20px;
  margin: 8px 0;
  box-sizing: border-box;
}

input[type="text"]:focus {
  background-color: $backColorNav;
  color: white;
}

ul {
  list-style: none;
}

li {
  list-style: none;
  margin-bottom: 2%;
}

#consolidateButtonStack {
  background-color: $backColorNav;
  border-radius: 20px;
  // display: block;
  //float: left;
  //margin-left: 50%;
}

#editCardButton_Positioning {
  float: left;
}

#infoCardMan {
  margin-top: 20%;
}

/// Card Flipping effect

.flipInfoCard {
  padding: 1%;
  margin-top: 25%;
  margin-bottom: 4%;
  // background-color: transparent;
  width: 250px;
  height: 150px;
  border-radius: 5px;
  perspective: 1000px; /* Remove this if you don't want the 3D effect */
}

/* This container is needed to position the front and back side */
.flipInfoCardInner {
  position: relative;
  width: 100%;
  height: 100%;
  border-radius: 5px;
  text-align: center;
  transition: transform 0.8s;
  transform-style: preserve-3d;
}

/* Do an horizontal flip when you move the mouse over the flip box container */
.flipInfoCard:hover .flipInfoCardInner {
  transform: rotateY(180deg);
}

/* Position the front and back side */
.flipInfoCardFront,
.flipInfoCardBack {
  position: absolute;
  width: 100%;
  height: 100%;
  border-radius: 5px;
  -webkit-backface-visibility: hidden; /* Safari */
  backface-visibility: hidden;
}

/* Style the front side (fallback if image is missing) */
.flipInfoCardFront {
  background-color: $mainColor;
  color: $backColorNav;
  font-weight: bold;
  box-shadow: 0 4px 8px 0 rgba(0, 0, 0, 0.2);
}

/* Style the back side */
.flipInfoCardBack {
  background-color: $contrastColor;
  color: whitesmoke;
  font-weight: bolder;
  transform: rotateY(180deg);
}

.editCardButtonStyle {
  background-color: $backColorNav;
  color: $contrastColor;
}

#editButtonMessage {
  color: $mainColor;
  margin-top: -35px;
}
