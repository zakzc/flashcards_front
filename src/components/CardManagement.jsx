import React, { Component } from "react";
import { Redirect } from "react-router-dom";
// Standard Log out page
import UserIsLoggedOut from "./views/UserIsLoggedOut";
import ManageCards from "./CardManagement/ManageCards";

export default class CardManagement extends Component {
  constructor(props) {
    super(props);
    this.state = {
      userIsLoggedIn: props.userIsLoggedIn,
      currentUser: props.currentUser,
      currentStack: props.currentStack,
      addOrManage: false,
    };
    // * Methods
    this.UserIsLoggedOut = UserIsLoggedOut;
  }

  ////////////////
  // View methods
  ////////////////

  cardManagementLayout() {
    return (
      <React.Fragment>
        <div id="cardManagementLayout">
          <ManageCards
            userIsLoggedIn={this.state.userIsLoggedIn}
            currentUser={this.state.currentUser}
            currentStack={this.state.currentStack}
          />
        </div>
      </React.Fragment>
    );
  }

  cardManagementView() {
    // console.log("CARD MANAGEMENT");
    return (
      <React.Fragment>
        <div id="cardManagementView">
          {this.state.userIsLoggedIn ? (
            <div id="loggedInView">{this.cardManagementLayout()}</div>
          ) : (
            <Redirect to="/login" />
          )}
        </div>
      </React.Fragment>
    );
  }

  render() {
    return (
      <React.Fragment>
        <div id="cardManagementView">{this.cardManagementView()}</div>
      </React.Fragment>
    );
  }
}
