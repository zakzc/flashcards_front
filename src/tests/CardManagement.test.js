import React from "react";
import Enzyme from "enzyme";
///
import Adapter from "enzyme-adapter-react-16";
import { shallow } from "enzyme";
///
import CardManagement from "../components/CardManagement";
import ManageCards from "../components/CardManagement/ManageCards";

Enzyme.configure({ adapter: new Adapter() });

// Parent

describe("\n __| CardManagement |__ ", () => {
  const cardManagement = shallow(<CardManagement />);

  describe("Initialization", () => {
    it("Matches snapshot", () => {
      expect(cardManagement).toMatchSnapshot();
    });

    it("Initializes the State", () => {
      expect(cardManagement.state().addOrManage).toEqual(false);
    });

    describe("Imports", () => {
      test("UserIsLoggedOut", () => {
        expect(cardManagement.find("UserIsLoggedOut")).toBeTruthy();
      });
      test("ManageCards", () => {
        expect(cardManagement.find("ManageCards")).toBeTruthy();
      });
    });
  });
});

// Child
describe("\n  ___| ManageCards |___ ", () => {
  const manageCards = shallow(<ManageCards />);

  describe("\tInitialization process", () => {
    it("Renders the ManageCards component", () => {
      expect(manageCards).toMatchSnapshot();
    });

    it("Initializes the State", () => {
      expect(manageCards.state().cardForEditing).toEqual("");
      expect(manageCards.state().cardForEditingId).toEqual("");
      expect(manageCards.state().tempNewFront).toEqual("");
      expect(manageCards.state().tempNewBack).toEqual("");
      expect(manageCards.state().newStack).toEqual("");
      expect(manageCards.state().editingMode).toEqual(false);
      expect(manageCards.state().updatedItemMessage).toEqual(false);
    });

    describe("\tImports", () => {
      test("updateCards", () => {
        expect(manageCards.find("updateCards")).toBeTruthy();
      });
      describe("\tSvg imports", () => {
        test("Consolidate", () => {
          expect(manageCards.find("Consolidate")).toBeTruthy();
        });
        test("EditCardIcon", () => {
          expect(manageCards.find("EditCardIcon")).toBeTruthy();
        });
      });
    });
  });
});
