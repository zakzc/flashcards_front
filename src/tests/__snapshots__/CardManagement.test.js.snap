// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`
 __| CardManagement |__  Initialization Matches snapshot 1`] = `ShallowWrapper {}`;

exports[`
- -> | ManageCards (child) |  	Initialization process Renders the ManageCards component 1`] = `ShallowWrapper {}`;
